### PR TITLE
Add Test Suite for Feedback 

### PR DESCRIPTION
**Pull request by shravanhariharan2**
Setup the infrastructure for feedback testing and for future test suites

### CODE REVIEW

File: tests/data/FactoryUtils.ts
Comment (shravanhariharan2) [inline tests/data/FactoryUtils.ts:12]: this function wasn't working for ranges with `scale = 1` (since any integer mod 1 is 0, so it would always return `minimum`), so I refactored it a bit

**Review by shravanhariharan2 — COMMENTED** [review]
@sumeet-bansal any thoughts on the whole setup? The idea is for a test suite for a particular API, the respective controller can be accessed as a singleton with services/repositories injected already. Will implement the rest of actual tests after making any setup changes needed.

File: tests/TestUtils.ts
Comment (sumeet-bansal) [inline tests/TestUtils.ts:2]: https://flaviocopes.com/how-to-sort-array-of-objects-by-property-javascript/

File: tests/TestUtils.ts
Comment (sumeet-bansal) [inline tests/TestUtils.ts:10]: https://jestjs.io/docs/en/expect#expectarraycontainingarray

File: tests/data/FactoryUtils.ts
Comment (sumeet-bansal) [reply tests/data/FactoryUtils.ts:12]: nit: rename `range` to `value` or something

File: tests/data/FeedbackFactory.ts
Comment (sumeet-bansal) [inline tests/data/FeedbackFactory.ts:11]: Should this should be `...sub`?

File: tests/data/FeedbackFactory.ts
Comment (sumeet-bansal) [inline tests/data/FeedbackFactory.ts:24]:
```suggestion
    return FactoryUtils.pickRandomValue(Object.values(FeedbackType));
```

File: tests/data/FactoryUtils.ts
Comment (sumeet-bansal) [inline tests/data/FactoryUtils.ts:13]:
```suggestion
    return (range % scale) * scale;
```

File: tests/controllers/index.ts
Comment (sumeet-bansal) [inline tests/controllers/index.ts:10]: Can we avoid interacting with the container entirely and manually construct the feedback service?

File: tests/TestUtils.ts
Comment (shravanhariharan2) [reply tests/TestUtils.ts:10]: I think this is slightly different - we need a way to have `expectArrayContaining` but with objects that partially match (`expectArrayContaining` expects exact matches).  In the case with feedback, we're comparing two arrays of `Feedback[]` and `PublicFeedback[]`, and in other tests, we'd compare the `*Model[]` array to `Public*[]` which also only partially match. There doesn't seem to be a way to do this through jest's current API

Comment (sumeet-bansal) [reply tests/TestUtils.ts:10]: Why not map the Model arrays to the PublicModel arrays?

File: tests/data/FactoryUtils.ts
Comment (shravanhariharan2) [reply tests/data/FactoryUtils.ts:13]: I'm not sure these are doing the same thing (with the new min, max, scale thing). Let's say we want a value between 30 and 120 with a scale of 30. `range` will give us a value between 30 and 120 that has scale of 1 (e.g. 68), and current line 13 will floor this value to the nearest `scale` (so 68 -> 60, 56 -> 30, 90 -> 90).

New line 13 would do: `(range % scale) * scale` = (68 % 30) * 30 = 8 * 30 = 240 which isn't in the range

File: tests/TestUtils.ts
Comment (shravanhariharan2) [reply tests/TestUtils.ts:2]: oh yeah we can do that for the other tests, but the feedback factory creates `Feedback` objects rather than `FeedbackModel` ones because `FeedbackModel` relies on a `UserModel` to submit it. 

I personally think it better mimics the actual flow of feedback to have `Feedback` objects created in the factory and `PortalState::submitFeedback(feedback: Feedback, user: UserModel)` actually attaching a user to it, but I suppose having  `FeedbackFactory::create(n: number, user: UserModel): FeedbackModel[]` instead doesn't hurt too

Comment (sumeet-bansal) [reply tests/TestUtils.ts:2]: Current FeedbackFactory's fine; I meant for the assertions, we can map Model[] to PublicModel[] instead of creating our own matcher. I think the tests should use the controller to submit feedback instead though.

Comment (shravanhariharan2) [reply tests/TestUtils.ts:2]: I was thinking about having a different test for each controller alone so that only one thing's tested at a time, so all data that would need to be inserted for a test is done with `PortalState` and just 1 controller is tested rather than 2+.

Comment (sumeet-bansal) [reply tests/TestUtils.ts:2]: > I was thinking about having a different test for each controller alone so that only one thing's tested at a time

Do you mean one test per route/controller method? Most of the POST/GET routes are pretty basic CRUD operations so I think those are fine to test together, e.g. submit a feedback and then get it. Looking at `FeedbackService` the only code that's executed in that flow is pretty simple stuff like creating and inserted via TypeORM and then a basic `SELECT` query.

**Comment by shravanhariharan2** [comment]
@sumeet-bansal A while back we talked about how we should test validation, and you said that we can do manual validation with class-validator for that. Doesn't that cause a lot of duplicate validation checks between the test suite and middleware though? This along with the need to implement our own DI with this current setup makes me wonder if actually spinning up the server and sending requests is better than just using the controller methods. Thoughts?

Also, since we were planning on exposing API wrapper functions as a package anyway, I think this would a good time to set that up for both testing and for the package. This would also keep the responsibility of DI / validation in the server's hands and not ours.

**Comment by sumeet-bansal** [comment]
> Doesn't that cause a lot of duplicate validation checks between the test suite and middleware though?

From what I remember, routing-controllers attaches its own middleware for validation, so I don't think any controllers we initialize validate inputs. If you want to test something, you should only test that one thing, which's why I said if we're testing validation we can use the class-validator `validate` call to only test the validators.

In general, I think it's better to avoid DI magic since it can be really confusing to understand how something that uses it works. Spinning up a server and sending requests works if we're testing the server as a black box, but if we want to make sure specific things are working as intended (e.g. validation, the logic in some service method, database queries) then we need to test those pieces in isolation.

Using the server as a black box also means the entire test server needs to be in a testable state since all the code runs instead of just the bits you're testing, so for example if you want to make sure a member can't access some object like feedback or a merch order, you need to create the db connection and server, get
 auth tokens,

**Comment by sumeet-bansal** [comment]
> Also, since we were planning on exposing API wrapper functions as a package anyway, I think this would a good time to set that up for both testing and for the package.

I'm waiting for this PR at least to go through since a bunch of files are gonna get moved around, but yeah, I'll get on that soon. I started a little on it and it doesn't seem too difficult.

**Comment by shravanhariharan2** [comment]
@sumeet-bansal in this setup, I try to keep the tests as small as possible by hiding away details like connection and portal state in separate factories, but there's some inconsistency in abstraction (such as calling `conn.manager.findOne()` inside tests) and some code duplication that I'm not sure how to change. In general, I'm still not really sure where to draw the line between clean abstraction and overcomplicating the architecture, so let me know what you think of the design and its sustainability.

Also feel free to suggest any additional test cases

**Comment by shravanhariharan2** [comment]
@sumeet-bansal can you re-review this when you get the chance?

File: types/HttpEnums.ts
Comment (sumeet-bansal) [inline types/HttpEnums.ts:1]: Is this used anywhere? I don't see any uses besides re-exporting in `types/Enums`.

File: tests/feedback.test.ts
Comment (sumeet-bansal) [inline tests/feedback.test.ts:50]: I think it'd be a little cleaner here if we used the GET route instead of querying the database, plus we get to make sure that route's working too.

File: tests/feedback.test.ts
Comment (sumeet-bansal) [inline tests/feedback.test.ts:62]: Same thing here with matching the entire object once instead of separate assertions for each property.

File: tests/feedback.test.ts
Comment (sumeet-bansal) [inline tests/feedback.test.ts:68]: nit: this was a little confusing to read, maybe something like "admins can feedback from any/all user, members can only see their own"

Comment (sumeet-bansal) [reply tests/feedback.test.ts:68]: might also be worth breaking up into separate tests: "admins can see all users' feedback" and "members can only see their feedback"

File: tests/feedback.test.ts
Comment (sumeet-bansal) [inline tests/feedback.test.ts:89]: nit: "admin can acknowledge and reward points for feedback"

File: tests/feedback.test.ts
Comment (sumeet-bansal) [inline tests/feedback.test.ts:106]: nit: "admin can ignore and not reward points for feedback"

File: tests/feedback.test.ts
Comment (sumeet-bansal) [inline tests/feedback.test.ts:128]: nit: `feedbackToAcknowledge`, `feedbackToIgnore`

File: tests/feedback.test.ts
Comment (sumeet-bansal) [inline tests/feedback.test.ts:137]: Why create single-use functions here and then call them right after? vs inlining them.

File: tests/index.ts
Comment (sumeet-bansal) [inline tests/index.ts:7]: Is this all necessary if we're manually instantiating everything through constructors?

File: tests/feedback.test.ts
Comment (shravanhariharan2) [reply tests/feedback.test.ts:137]: Mainly for conveying via. function name what the function is doing, but with the variable renaming change above, these functions aren't needed anymore.

File: tests/index.ts
Comment (shravanhariharan2) [reply tests/index.ts:7]: Don't think it's necessary anymore

File: tests/feedback.test.ts
Comment (shravanhariharan2) [reply tests/feedback.test.ts:27]: Wouldn't that be an extra unnecessary query though? Also, this test is explicitly for checking the `submitFeedback` response object's validity, and if we want to check the response of `getFeedback`, then I think that should be in a different test

File: tests/controllers/FeedbackControllerMock.ts
Comment (shravanhariharan2) [reply tests/controllers/FeedbackControllerMock.ts:7]: I mainly built it to abstract away the database connection `conn` (and things like `PortalState`) from the tests to make them simpler and easier to read

File: tests/feedback.test.ts
Comment (sumeet-bansal) [reply tests/feedback.test.ts:27]: I just don't get the point of the test. Like, the response object is something we return and we have a type for it so why do we need to check its validity. If we want to make sure the submission went through, then we could call GET or query the database, but I don't see the point in checking the response object since it's basically just whatever we put in the request plus some constants.

File: tests/feedback.test.ts
Comment (shravanhariharan2) [reply tests/feedback.test.ts:50]: Wouldn't that make this test case dependent on whether or not a different controller is working? If that controller wasn't working, then this test case would fail, which might cause some confusion / extra looking since the actual feedback functionality would be working

File: tests/feedback.test.ts
Comment (shravanhariharan2) [reply tests/feedback.test.ts:31]: See comment above about purpose of FeedbackControllerMock

File: tests/feedback.test.ts
Comment (sumeet-bansal) [reply tests/feedback.test.ts:53]: The GET's a pretty simple route, just a permissions check and db query.

File: tests/controllers/FeedbackControllerMock.ts
Comment (sumeet-bansal) [reply tests/controllers/FeedbackControllerMock.ts:7]: The `beforeAll/beforeEach` already take care of the db connection, so it's
```typescript
await FeedbackControllerMock.mock(users);
```
vs.
```typescript
await new PortalState().createUsers(users).write();
```
At that point you basically have a whole wrapper for that change alone, and you'd have to do something similar for all the other controllers. Makes more sense to just keep the `PortalState` stuff in the test itself since that's what it was made for. Also as a nit, nothing's being "mocked"; we're creating real users in the database and letting the controller execute normally instead of mocking behavior for whatever components are in the controller.

**Comment by shravanhariharan2** [comment]
@sumeet-bansal here's what it looks like without the FeedbackControllerMock class. IMO the tests themselves are a little less clean since now there's a tiny bit more setup involved within the connection and the controller parameters are inconsistently nested (i.e. some params such as feedback are wrapped while others such as user are not). Plus, the tests have different layers of abstraction now with manually writing portal state and calling top-level controllers, which may make it slightly more confusing to read.

I personally think it's worth wrapping each controller in an extra class that manages connections and cleans up requests/responses, but maybe there's a better way to design it, or maybe I'm too critical on having small tests and in reality it doesn't really matter. Let me know what you think

File: tests/controllers/ControllerFactory.ts
Comment (sumeet-bansal) [inline tests/controllers/ControllerFactory.ts:6]: `DatabaseConnection` has a class-wide connection because it makes managing connections simpler like with closing them out, but it's fine for `ControllerFactory` to instantiate the controller  every time.

File: tests/feedback.test.ts
Comment (sumeet-bansal) [inline tests/feedback.test.ts:35]:
```suggestion
    await feedbackController.submitFeedback({ feedback }, user);
```

File: tests/feedback.test.ts
Comment (sumeet-bansal) [inline tests/feedback.test.ts:58]: Doesn't the sample test already verify that things get persisted to the database?

File: tests/feedback.test.ts
Comment (sumeet-bansal) [inline tests/feedback.test.ts:93]: I still don't understand the need to verify the submission responses.

File: tests/feedback.test.ts
Comment (sumeet-bansal) [inline tests/feedback.test.ts:134]: Can we do this via the user controller? I think it's fine to use the simple GETs for different controllers.

File: tests/feedback.test.ts
Comment (sumeet-bansal) [inline tests/feedback.test.ts:154]: Same here about simple GETs vs direct queries.

File: tests/feedback.test.ts
Comment (sumeet-bansal) [inline tests/feedback.test.ts:171]: nit: these can be inlined

File: types/HttpEnums.ts
Comment (sumeet-bansal) [reply types/HttpEnums.ts:1]: ^

**Review by sumeet-bansal — COMMENTED** [review]
The thing about wrapped responses makes sense but IMO it's pretty minor/not worth the tradeoff of creating wrappers for every controller, and I think a test for a controller should deal directly with the controller's full response.

For the connection thing, `ControllerFactory` and `PortalState::write` could call `DatabaseConnection` themselves instead of requiring a connection as a parameter, but PortalStates were made to be used in tests so I don't think it's a problem that they're being used in tests. They were made to be flexible and to accommodate whatever scenario a test might need to setup, and I'm sure other tests are going to use them in weird and niche ways, so abstracting them away behind rigid functions for all possible test cases seem unsustainable: merch tests would require their own function for writing users and store things, sliding leaderboard tests would require their own function for writing users and events and attendances, etc. It'd be simpler and more readable to directly define the PortalState in the test.

File: tests/feedback.test.ts
Comment (shravanhariharan2) [reply tests/feedback.test.ts:93]: Yeah you're right, this assertion is pretty redundant here

File: tests/feedback.test.ts
Comment (shravanhariharan2) [reply tests/feedback.test.ts:58]: Yup - I think what we want to assert instead is that the scope/type are correct for feedback submission

Comment (shravanhariharan2) [reply tests/feedback.test.ts:58]: Also, slightly unrelated, do we want `PublicActivity` to return the `scope` as well? Since `PublicActivity` can now be either private or public depending on if it's fetched for the current user or a different one. Also we should probably rename `PublicActivity` since it's technically incorrect but still conventional since everything returned to the client is considered "public"

Comment (sumeet-bansal) [reply tests/feedback.test.ts:58]: I don't think `PublicActivity` needs to return `scope` since permissions are internal, but I don't have a strong opinion about it. The type name's probably fine to leave as is.

File: tests/controllers/ControllerFactory.ts
Comment (shravanhariharan2) [reply tests/controllers/ControllerFactory.ts:11]: I feel like it's alright to leave the controllers as singletons, just so that service/controller objects don't have to be re-created again every time a controller is invoked, even though that overhead is fairly small.

File: api/controllers/FeedbackController.ts
Comment (sumeet-bansal) [inline api/controllers/FeedbackController.ts:18]: We don't need to inject the service in the constructor? I see in the typedi and routing-controllers docs that we don't need the `@Inject` if the class has the `@Service` annotation, but I don't see anything like that for controller annotations.

File: tests/feedback.test.ts
Comment (sumeet-bansal) [inline tests/feedback.test.ts:55]:
```suggestion
  test('has proper activity scope and type', async () => {
```

File: tests/feedback.test.ts
Comment (sumeet-bansal) [inline tests/feedback.test.ts:84]: Can we do a `toStrictEqual([...])` here (and in the other array assertions) or do we need to use the `arrayContaining`?

File: tests/controllers/ControllerFactory.ts
Comment (sumeet-bansal) [inline tests/controllers/ControllerFactory.ts:16]: As an aside, we should prob pass in the S3 config to the `StorageService` constructor instead of importing.

**Review by sumeet-bansal — APPROVED** [review]

**Comment by sumeet-bansal** [comment]
Approved but @michl1001 should review it too before merging.

**Comment by michl1001** [comment]
I looked through the changes, _haven't tested it on my machine yet_ but so far I can't see anything wrong, in the comments for `feedback.test.ts`, Sumeet made a good point on checking arrays and idk if it was ever addressed. Orher than that, everything looks in order.

File: api/controllers/FeedbackController.ts
Comment (shravanhariharan2) [reply api/controllers/FeedbackController.ts:18]: Looks like we don't need them - in fact, including them throws an error when sending the controller a request:
`ServiceNotFoundError: Cannot determine a class of the requesting service \"undefined\"`

File: tests/feedback.test.ts
Comment (shravanhariharan2) [reply tests/feedback.test.ts:84]: Nope don't think so, since the database tables have indices on uuid, they are pulled in sorted order by uuid which is unpredictable (since feedback factory doesn't produce models)

File: tests/controllers/ControllerFactory.ts
Comment (shravanhariharan2) [reply tests/controllers/ControllerFactory.ts:16]: I personally think it's fine as is, just so that different services/controllers don't need to know about lower-level aws configs when instantiating the storage service, and so that aws doesn't need to get imported everywhere

**Pull request merged**
